src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp

```diff
@@ -0,0 +1,167 @@
+    auto group_by_list_changed = false;
+
+    // Main loop. Iterate over the tables and its group-by columns, get table constraints, and try to reduce for each
+    // constraint.
+    for (const auto& [stored_table_node, group_by_columns] : group_by_columns_per_table) {
+      auto unique_columns = std::set<ColumnID>();
+
+      const auto& table = Hyrise::get().storage_manager.get_table(stored_table_node->table_name);
+      const auto& table_constraints = table->get_soft_unique_constraints();
+      if (table_constraints.empty()) {
+        // early exit for current table if no constraints are set
+        continue;
+      }
+
+      // Gather sizes of table constraints (i.e., the number of columns) to start with the shortest constraint (either
+      // unique or primary key) in hope that the shorter one will later form the group-by clause.
+      std::vector<std::pair<size_t, size_t>> constraints_position_and_size;
+      constraints_position_and_size.reserve(table_constraints.size());
+      auto constraint_id = size_t{0};
+      for (const auto& table_constraint : table_constraints) {
+        constraints_position_and_size.emplace_back(constraint_id, table_constraint.columns.size());
+        ++constraint_id;
+      }
+      std::sort(
+          constraints_position_and_size.begin(), constraints_position_and_size.end(),
+          [](const auto& constraint_1, const auto& constraint_2) { return constraint_1.second < constraint_2.second; });
+
+      // Try to reduce the group-by list one constraint at a time, starting with the shortest constraint. As soon as
+      // one reduction took place, we can ignore the remaining constraints.
+      for (const auto& [position, size] : constraints_position_and_size) {
+        auto const& table_constraint = table_constraints[position];
+        group_by_list_changed |=
+            reduce_for_constraint(table_constraint, group_by_columns, stored_table_node, aggregate_node);
+        if (group_by_list_changed) break;
+      }
+    }
+
+    // In case the initial query plan root returned the same columns in the same column order and was not a projection,
+    // it is likely that the result of the current aggregate was either the root itself or only operators followed that
+    // do not modify the column order (e.g., sort or limit). In this case, we need to restore the initial column order
+    // by adding a projection with the initial column_references since we changed the column order by moving columns
+    // from the group-by list to the aggregations.
+    if (group_by_list_changed && initial_aggregate_column_expressions == root_column_expressions &&
+        lqp->type != LQPNodeType::Projection) {
+      const auto projection_node = std::make_shared<ProjectionNode>(root_column_expressions);
+      lqp_insert_node(lqp, LQPInputSide::Left, projection_node);
+    }
+
+    return LQPVisitation::VisitInputs;
+  });
+}
+
+bool DependentGroupByReductionRule::reduce_for_constraint(
```

Comment:
Changed to `reduce_group_by_columns_for_constraint` and added the following comment.

```
/**
 * This function reduces the group-by columns of @param aggregate_node for the inputs @param stored_table_node and
 * @param table_constraint. @param group_by_columns is passed as it already stores the ColumnID's which are used for
 * intersection instead of retrieving them from @param aggregate_node again.
 *
 * @returns   Boolean value denoting whether at the group-by list of @param aggregate_node changed.
 */
```